packages/block-library/src/heading/heading-level-dropdown.js

```diff
@@ -0,0 +1,115 @@
+/**
+ * WordPress dependencies
+ */
+import {
+	Button,
+	Dropdown,
+	Toolbar,
+	ToolbarGroup,
+	__experimentalToolbarItem as ToolbarItem,
+} from '@wordpress/components';
+import { __, sprintf } from '@wordpress/i18n';
+import { DOWN } from '@wordpress/keycodes';
+
+/**
+ * Internal dependencies
+ */
+import HeadingLevelIcon from './heading-level-icon';
+
+const HEADING_LEVELS = [ 1, 2, 3, 4, 5, 6 ];
+
+const POPOVER_PROPS = {
+	className: 'block-library-heading-level-dropdown',
+	isAlternate: true,
+};
+
+/** @typedef {import('@wordpress/element').WPComponent} WPComponent */
+
+/**
+ * HeadingLevelDropdown props.
+ *
+ * @typedef WPHeadingLevelDropdownProps
+ *
+ * @property {number}                 selectedLevel The chosen heading level.
+ * @property {(newValue:number)=>any} onChange      Callback to run when
+ *                                                  toolbar value is changed.
+ */
+
+/**
+ * Dropdown for selecting a heading level (1 through 6).
+ *
+ * @param {WPHeadingLevelDropdownProps} props Component props.
+ *
+ * @return {WPComponent} The toolbar.
+ */
+export default function HeadingLevelDropdown( { selectedLevel, onChange } ) {
+	return (
+		<ToolbarItem>
+			{ ( itemProps ) => (
+				<Dropdown
+					popoverProps={ POPOVER_PROPS }
+					renderToggle={ ( { onToggle, isOpen } ) => {
+						const openOnArrowDown = ( event ) => {
+							if ( ! isOpen && event.keyCode === DOWN ) {
+								event.preventDefault();
+								event.stopPropagation();
+								onToggle();
+							}
+						};
```

Comment:
I noticed that pressing arrow keys to move between the block toolbar buttons isn't working when this dropdown toggle button has focus, and I guess that's because this `openOnArrowDown` is overwriting `itemProps.onKeyDown`.

To fix that, you'll have to call `itemProps.onKeyDown` here like the suggestion below. But I wonder if just using `ToolbarButton` instead of `ToolbarItem` + `Button` wouldn't be better.
```suggestion
						const openOnArrowDown = ( event ) => {
							if ( itemProps.onKeyDown ) {
						        	itemProps.onKeyDown( event );
							}
							if ( ! isOpen && event.keyCode === DOWN ) {
								event.preventDefault();
								event.stopPropagation();
								onToggle();
							}
						};
```